### PR TITLE
Pal 919 topic offset actor system

### PR DESCRIPTION
**Pull request by dev930018**
* removed spring web in preparation for later ticket
* added topic offset actor system

### CODE REVIEW

**Review by pd104923 — DISMISSED** [review]

**Review by dbz120896 — DISMISSED** [review]

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/config/ApplicationConfiguration.java [hunk 1]
```diff
@@ -78,6 +81,21 @@ ErrorHandlingService loggingErrorHandler() {
         return (String token, FilteredResourceRequest request, Throwable error) -> LOGGER.error("Token {} and request {} threw exception {}", token, request, error.getMessage());
     }
 
+    /**
+     * Create the TokenOffsetActorSsytem, controlling get access to the persistence layer and asynchronously blocking until offsets
```
Comment (dev-958) [inline filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/config/ApplicationConfiguration.java]: If you are blocking, you are not doing reactive, you are not doing akka. Are you sure this statement is correct?

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+import java.time.temporal.ChronoUnit;
+import java.util.HashMap;
+import java.util.Optional;
+
+/**
+ * Manages getting from the persistence layer such that requests are 'blocked' until an offset arrives.
+ * This aims to remove any long-polling of the persistence store (redis) while ensuring offsets from
+ * kafka can be set before -or- after they are got by the client's websocket handler.
+ */
+public final class TokenOffsetActorSystem extends AbstractBehavior<TokenOffsetActorSystem.TokenOffsetCmd> {
```
Comment (dev-958) [inline filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java]: This is a really terrible name. Having "ActorSystem" in it makes it seem like a subclass of the akka ActorSystem - which it is not. In programming terms it's a controller.

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    // Timeout on the ActorFlow::ask
+    // The unit can't be longer than DAYS (anything more is an estimated duration)
+    // Despite this, we can just put the max value akka will accept, which seems to be about 248 days
+    private static final Duration TIMEOUT = Duration.of(Integer.MAX_VALUE / 100, ChronoUnit.SECONDS);
+
+    private final TokenOffsetPersistenceLayer persistenceLayer;
+    // Map from Tokens to TokenOffsetWorker actors
+    private final HashMap<String, ActorRef<WorkerCmd>> inFlightWorkers = new HashMap<>();
```
Comment (dev-958) [inline filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java]: just "workers" would be fine, yes?

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+     *    source
+     *        .mapAsync((token, offset) -> persistenceLayer.putOffset(token, offset)
+     *                .thenApply(ignored -> (token, offset)))
+     *        .to(TokenOffsetActorSystem.asSetterSink(tokenOffsetActor))
+     *        .run(actorSystem);
+     * }</pre>
+     */
+    public static Sink<Pair<String, Long>, NotUsed> asSetterSink(final ActorRef<TokenOffsetCmd> tokenOffsetActor) {
+        return ActorFlow.ask(tokenOffsetActor, TIMEOUT, TokenOffsetCmd.AckTellWorker::new)
+                // Easier to do ActorFlow().to(Sink.ignore) than ActorSink()
```
Comment (dev-958) [inline filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java]: let this be our secret, eh?

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+     *
+     * @param tokenOffsetActor an {@link ActorRef} to the {@link ActorSystem} created by the
+     *                         {@link TokenOffsetActorSystem#create(TokenOffsetPersistenceLayer)} method
+     * @return a {@link Flow} from tokens to {@link Pair}s of tokens and their offsets
+     * @implNote The {@link Flow} remains strictly ordered, so a long-blocking first element will delay all subsequent
+     * elements.
+     */
+    public static Flow<String, Pair<String, Long>, NotUsed> asGetterFlow(final ActorRef<TokenOffsetCmd> tokenOffsetActor) {
+        return ActorFlow.ask(tokenOffsetActor, TIMEOUT, TokenOffsetCmd.SpawnWorker::new)
+                // Downcast SetOffset to Pair<String, Long>
```
Comment (dev-958) [inline filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java]: not strictly a downcast

**Review by dev-958 — CHANGES_REQUESTED** [review]
Seems to be no expectation that persistence throws exceptions. Extra message type and state required now, add the reporting at a later stage when we put it in for Kafka handling.

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java (hunk 2 above)
Comment (dev930018) [reply filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java]: The object produced by the `create` method, which is the only "constructor" in the class, is an ActorSystem, even though the class is an implementation of Behaviors
Happy to rename if you think Controller is better

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java (hunk 4 above)
Comment (dev930018) [reply filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java:134]: This used to be a downcast from `class AckTellWorker implements Pair<A, B>` to `class Pair<A, B>`, and the code was just `.map(x -> x)`. Removed as irrelevant

File: filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java (hunk 5 above)
Comment (dev930018) [reply filtered-resource-service/src/main/java/uk/gov/gchq/palisade/service/filteredresource/repository/TokenOffsetActorSystem.java]: This used to be a downcast from `class AckTellWorker implements Pair<A, B>` to `class Pair<A, B>`, and the code was just `.map(x -> x)`
Removed as irrelevant

**Review by dev-958 — APPROVED** [review]
Excellent

**Pull request merged**